orderbook/src/orderbook.rs

```diff
@@ -85,6 +89,52 @@ impl OrderBook {
         orders.retain(|_, order| has_future_valid_to(now_in_epoch_seconds, &order.order_creation));
     }
 
+    async fn remove_settled_orders(&self, settlement_contract: &GPv2Settlement) {
+        let uids_to_be_removed = self.get_uids_of_settled_orders(settlement_contract).await;
+        info!(
+            "removing following settled orders from orderbook: {:?}",
+            uids_to_be_removed
+        );
+        let mut orders = self.orders.write().await;
+        orders.retain(|uid, _| uids_to_be_removed.contains(uid));
+    }
+
+    async fn get_uids_of_settled_orders(
+        &self,
+        settlement_contract: &GPv2Settlement,
+    ) -> Vec<OrderUid> {
+        let orders = self.orders.read().await;
+        let uid_futures = orders.iter().map(|(uid, _)| async move {
+            (
+                *uid,
+                self.has_not_yet_been_settled(*uid, settlement_contract)
+                    .await,
+            )
+        });
+        let uid_pairs: Vec<(OrderUid, bool)> = join_all(uid_futures).await;
+        uid_pairs
+            .iter()
+            .filter(|(_, not_yet_settled)| *not_yet_settled)
+            .map(|(uid, _)| *uid)
+            .collect()
+    }
+
+    async fn has_not_yet_been_settled(
+        &self,
+        uid: OrderUid,
+        settlement_contract: &GPv2Settlement,
+    ) -> bool {
+        let filled_amount = settlement_contract
+            .filled_amount(uid.0.to_vec())
+            .call()
+            .await
+            .unwrap_or(U256::MAX);
```

Comment:
In order to test this, this call would have to be decoupled from the orderbook class. You could either achieve this by instead of passing GPv2Settlement in the constructor you pass a `OrderFillmentQuerying` trait which exposes `filled_amount(order_id)`. In tests you can mock this to your liking, in prod you implement a struct that forwards this call to an inner `GPv2Settlement`.

Instead of passing a trait (which I think is more object oriented, but functionally no different as long as there is just one call you intend to mock) you could also pass ing a `filled_amount` function into the constructor (@e00E usually prefers this approach). In prod this function captures GPv2Settlement and forwards the call, in test you create a local closure just returning the amounts you want.

Reply:
Since we have easy to run integration tests now you could add a check for filled orders being removed to our existing test. I like how when doing it like that we don't need to adapt the code under test.

Reply:
I like both suggestions. THanks guys!

I will pick one of Felix's suggestion to close the PR and then I will try to add Valentin's option additional with a new PR to the e2e test, if this is okay for you guys.

Reply:
That's okay but the way I imagined it is that the e2e obsoletes the task Felix suggested so then implementing that first would be wasted.

Reply:
I kinda like to test functions without e2e tests, as this gives us quicker feedback during development processes. I always think of the F>I>R>S>t principle.  Though, yeah, in this particular instance, it might make sense to just have the e2e test.

I coded the e2e test in the newest commit.